From 664fd0d3d8860120b57e1739e33b6786a725c0d6 Mon Sep 17 00:00:00 2001
From: Devi Priya <quic_devipriy@quicinc.com>
Date: Sun, 17 Dec 2023 02:37:44 +0530
Subject: [PATCH] drivers: soc: qcom: Add support to load GENI SE firmware

Add support to program the GENI IMAGE CFG registers, configure clock,
output control, FW revision and protocol registers.
While at it, Add support to program the GENI RAM Address space(CFG_RAMn)
for Firmware based SEs.

Change-Id: If8004095dd264c53a10167f2a90ec5d1bf5e2120
Signed-off-by: Devi Priya <quic_devipriy@quicinc.com>
---
 drivers/soc/qcom/Kconfig              |   7 +
 drivers/soc/qcom/qcom-geni-se.c       | 384 ++++++++++++++++++++++++++
 drivers/spi/spi-geni-qcom.c           |   4 +
 drivers/tty/serial/qcom_geni_serial.c |   5 +
 include/linux/soc/qcom/geni-se.h      |  14 +
 5 files changed, 414 insertions(+)

diff --git a/drivers/soc/qcom/Kconfig b/drivers/soc/qcom/Kconfig
index 6916ac5bd2be..2842f1320e14 100644
--- a/drivers/soc/qcom/Kconfig
+++ b/drivers/soc/qcom/Kconfig
@@ -51,6 +51,13 @@ config QCOM_GENI_SE
 	  driver is also used to manage the common aspects of multiple Serial
 	  Engines present in the QUP.
 
+
+config QCOM_GENI_SE_FW_LOAD
+	bool "Qualcomm GENI Serial Engine Firmware Loader Support"
+	depends on ARCH_QCOM || COMPILE_TEST
+	help
+	  Say Y here to enable support to load GENI SE FW
+
 config QCOM_GSBI
 	tristate "QCOM General Serial Bus Interface"
 	depends on ARCH_QCOM || COMPILE_TEST
diff --git a/drivers/soc/qcom/qcom-geni-se.c b/drivers/soc/qcom/qcom-geni-se.c
index ba788762835f..bd238d0f9605 100644
--- a/drivers/soc/qcom/qcom-geni-se.c
+++ b/drivers/soc/qcom/qcom-geni-se.c
@@ -891,6 +891,390 @@ int geni_icc_disable(struct geni_se *se)
 }
 EXPORT_SYMBOL(geni_icc_disable);
 
+#ifdef CONFIG_QCOM_GENI_SE_FW_LOAD
+
+#define SE_GENI_INIT_CFG_REVISION_OFFSET	0x0000
+#define SE_GENI_S_INIT_CFG_REVISION_OFFSET	0x0004
+#define SE_GENI_FORCE_DEFAULT_REG_OFFSET	0x0020
+#define SE_GENI_OUTPUT_CTRL_OFFSET		0x0024
+#define SE_GENI_CGC_CTRL_OFFSET			0x0028
+#define SE_GENI4_IMAGE_REG_OFFSET		0x0100
+#define SE_HW_PARAM_2_OFFSET			0x0E2C
+#define SE_GENI_FW_REVISION_OFFSET		0x1000
+#define SE_GENI_S_FW_REVISION_OFFSET		0x1004
+#define SE_GENI_CFG_RAMn_OFFSET			0x1010
+#define SE_GENI_CLK_CTRL_OFFSET			0x2000
+#define SE_DMA_IF_EN_OFFSET			0x2004
+#define SE_GENI_USE_MINICORES_MASK		BIT(12)
+
+typedef struct __attribute__ ((packed)) se_fw_hdr
+{
+    uint32_t firmware_magic;
+    /* Structure version number */
+    uint32_t se_fw_hdr_version;
+    /* QUPV3_HW_VERSION */
+    uint32_t qup_core_version;
+    /* Programmed into GENI_FW_REVISION */
+    uint16_t serial_protocol;
+    /* Programmed into GENI_FW_REVISION */
+    uint16_t firmware_version;
+    /* Programmed into GENI_INIT_CFG_REVISION */
+    uint16_t config_version;
+    /* Number of (uint32) GENI_FW_RAM words */
+    uint16_t firmware_sz_in_items;
+    /* Byte offset of GENI_FW_RAM array */
+    uint16_t firmware_offset;
+    /* Number of GENI_FW_CFG index/value pairs */
+    uint16_t config_sz_in_items;
+    /* Byte offset of GENI_FW_CFG indexes array */
+    uint16_t config_idx_offset;
+    /* Byte offset of GENI_FW_CFG values array */
+    uint16_t config_val_offset;
+} se_fw_hdr;
+
+static uint8_t uart_fw[] __attribute__ ((aligned (4))) = {
+    0x53, 0x45, 0x46, 0x57, 0x01, 0x00, 0x00, 0x00, 0x00, 0x00, 0x0a, 0x30, 0x02, 0x00, 0x04, 0x00,
+    0x0a, 0x00, 0xc6, 0x01, 0x1c, 0x00, 0x45, 0x00, 0x48, 0x08, 0x34, 0x07, 0x32, 0x00, 0x00, 0x00,
+    0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x28, 0x0f, 0x00,
+    0x20, 0x00, 0x78, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+    0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x42, 0x00, 0xc8, 0xc6, 0x7f, 0x00, 0x00, 0x00, 0x3f, 0x00,
+    0xc8, 0xc6, 0x7f, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+    0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+    0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+    0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+    0x00, 0x00, 0x00, 0x00, 0x21, 0x38, 0x7f, 0x00, 0xa1, 0x0b, 0x7f, 0x00, 0x82, 0x38, 0x7f, 0x00,
+    0x49, 0xd0, 0x9f, 0x00, 0x0d, 0x2e, 0x7f, 0x00, 0x63, 0x78, 0x9f, 0x00, 0x4f, 0xd0, 0x9f, 0x00,
+    0x15, 0x2e, 0x7f, 0x00, 0x63, 0x78, 0x9f, 0x00, 0x57, 0xd0, 0x9f, 0x00, 0x45, 0x10, 0x99, 0x00,
+    0x1d, 0x2e, 0x7f, 0x00, 0x63, 0x78, 0x9f, 0x00, 0x5f, 0x50, 0x9f, 0x00, 0x4b, 0x10, 0x99, 0x00,
+    0x25, 0x2e, 0x7f, 0x00, 0x63, 0x78, 0x9f, 0x00, 0x53, 0x10, 0x99, 0x00, 0x2d, 0x2e, 0x7f, 0x00,
+    0x6d, 0x10, 0x99, 0x00, 0x7f, 0x48, 0x9f, 0x00, 0xe7, 0x00, 0x9f, 0x00, 0x67, 0x18, 0x9f, 0x00,
+    0x7f, 0x78, 0x9f, 0x00, 0xe7, 0x00, 0x9f, 0x00, 0x6d, 0x60, 0x9f, 0x00, 0x03, 0x3a, 0x7f, 0x00,
+    0x79, 0x48, 0x9f, 0x00, 0x6c, 0x18, 0x9f, 0x00, 0x7f, 0x78, 0x9f, 0x00, 0xe7, 0x00, 0x9f, 0x00,
+    0x7f, 0x18, 0x9f, 0x00, 0x79, 0x68, 0x9f, 0x00, 0xa1, 0x0b, 0x7f, 0x00, 0xe7, 0x00, 0x9f, 0x00,
+    0x83, 0x60, 0x9f, 0x00, 0x28, 0x34, 0x7f, 0x00, 0x00, 0x08, 0x1f, 0x00, 0x04, 0x34, 0x7f, 0x00,
+    0x07, 0x39, 0x7f, 0x00, 0x20, 0x38, 0x7f, 0x00, 0x8f, 0x60, 0x9f, 0x00, 0x00, 0x80, 0x04, 0x00,
+    0x07, 0x39, 0x7f, 0x00, 0x8f, 0x50, 0x9f, 0x00, 0xed, 0x08, 0x9f, 0x00, 0x99, 0x60, 0x9f, 0x00,
+    0x08, 0x34, 0x7f, 0x00, 0xf5, 0x10, 0x99, 0x00, 0x08, 0x39, 0x7f, 0x00, 0x80, 0x1c, 0x7f, 0x00,
+    0xb7, 0x70, 0x9f, 0x00, 0xad, 0x00, 0x9a, 0x00, 0x02, 0x1c, 0x7f, 0x00, 0xd9, 0x58, 0x9f, 0x00,
+    0xc5, 0x78, 0x9f, 0x00, 0xd9, 0x10, 0x9f, 0x00, 0x07, 0x20, 0x7f, 0x00, 0xe7, 0x00, 0x9f, 0x00,
+    0xb1, 0x98, 0x98, 0x00, 0xd9, 0x78, 0x9f, 0x00, 0xd8, 0x30, 0x9f, 0x00, 0xc4, 0x10, 0x9f, 0x00,
+    0xd9, 0x78, 0x9f, 0x00, 0xe7, 0x00, 0x9f, 0x00, 0xd1, 0x18, 0x9f, 0x00, 0xbd, 0x68, 0x9f, 0x00,
+    0xd1, 0x78, 0x9f, 0x00, 0xc5, 0x60, 0x9f, 0x00, 0xd0, 0x10, 0x99, 0x00, 0x03, 0x3a, 0x7f, 0x00,
+    0xe7, 0x00, 0x9f, 0x00, 0xbd, 0x48, 0x9f, 0x00, 0xca, 0x18, 0x9f, 0x00, 0x20, 0x1c, 0x7f, 0x00,
+    0x28, 0x34, 0x7f, 0x00, 0x00, 0x00, 0x1f, 0x00, 0x00, 0x40, 0x1f, 0x00, 0xe7, 0x00, 0x9f, 0x00,
+    0xe0, 0x60, 0x9f, 0x00, 0xd8, 0x10, 0x9f, 0x00, 0x63, 0x78, 0x9f, 0x00, 0xdc, 0x18, 0x9f, 0x00,
+    0x28, 0x34, 0x7f, 0x00, 0xdd, 0x78, 0x9f, 0x00, 0x01, 0x60, 0x7f, 0x00, 0xe9, 0x60, 0x9f, 0x00,
+    0xd3, 0x78, 0x9f, 0x00, 0xed, 0x60, 0x9f, 0x00, 0x00, 0x00, 0x05, 0x00, 0x07, 0x39, 0x7f, 0x00,
+    0x99, 0x78, 0x9f, 0x00, 0x07, 0x39, 0x7f, 0x00, 0xf7, 0x60, 0x9f, 0x00, 0x18, 0x34, 0x7f, 0x00,
+    0x9f, 0x78, 0x9f, 0x00, 0x00, 0x00, 0x0f, 0x00, 0xe1, 0x0b, 0x7f, 0x00, 0x05, 0x01, 0x9f, 0x00,
+    0x00, 0x40, 0x1f, 0x00, 0x01, 0x60, 0x7f, 0x00, 0x03, 0x79, 0x9f, 0x00, 0x00, 0x00, 0x0f, 0x00,
+    0xa0, 0x04, 0x7f, 0x00, 0x0d, 0xd1, 0x9f, 0x00, 0x00, 0x40, 0x1f, 0x00, 0x00, 0x00, 0x00, 0x00,
+    0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+    0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+    0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+    0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x22, 0x00, 0x00, 0x00,
+    0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0xc8, 0x0c, 0x00,
+    0x10, 0x00, 0x5e, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+    0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x93, 0x00,
+    0x08, 0x80, 0x95, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+    0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+    0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+    0x01, 0x7a, 0x7f, 0x00, 0x08, 0x4c, 0x7f, 0x00, 0x01, 0x40, 0x7f, 0x00, 0xf7, 0x79, 0x9d, 0x00,
+    0xe9, 0x79, 0x9d, 0x00, 0x41, 0x88, 0x98, 0x00, 0x00, 0x00, 0x0e, 0x00, 0x01, 0x50, 0x7f, 0x00,
+    0x05, 0x20, 0x7f, 0x00, 0x04, 0x19, 0x9f, 0x00, 0x31, 0x8a, 0x98, 0x00, 0xd3, 0x49, 0x9f, 0x00,
+    0x01, 0x44, 0x6a, 0x00, 0x00, 0x80, 0x0b, 0x00, 0x48, 0xc2, 0x98, 0x00, 0x59, 0xb0, 0x98, 0x00,
+    0x43, 0x38, 0x7f, 0x00, 0x03, 0x20, 0x7f, 0x00, 0x04, 0x20, 0x7f, 0x00, 0x66, 0x40, 0x98, 0x00,
+    0x62, 0x40, 0x99, 0x00, 0x08, 0x1c, 0x7f, 0x00, 0x10, 0x1c, 0x7f, 0x00, 0x67, 0x78, 0x9f, 0x00,
+    0x00, 0x1e, 0x7f, 0x00, 0x03, 0x20, 0x7f, 0x00, 0x6a, 0x38, 0x99, 0x00, 0x04, 0x1c, 0x7f, 0x00,
+    0x6b, 0x68, 0x9f, 0x00, 0x07, 0x39, 0x68, 0x00, 0x20, 0x38, 0x7f, 0x00, 0x82, 0xb8, 0x7f, 0x00,
+    0x91, 0x50, 0x9f, 0x00, 0x1d, 0x41, 0x9f, 0x00, 0x05, 0x39, 0x7f, 0x00, 0x57, 0x11, 0x9f, 0x00,
+    0x7b, 0x60, 0x9f, 0x00, 0x01, 0x50, 0x7f, 0x00, 0x9d, 0x48, 0x8b, 0x00, 0x07, 0x39, 0x7f, 0x00,
+    0x87, 0x18, 0x9f, 0x00, 0x08, 0x60, 0x7f, 0x00, 0x08, 0x20, 0x7f, 0x00, 0x02, 0x24, 0x7f, 0x00,
+    0x8e, 0x29, 0x9a, 0x00, 0x48, 0x4c, 0x7f, 0x00, 0x8f, 0x79, 0x1f, 0x01, 0x95, 0x88, 0x98, 0x00,
+    0xdf, 0x79, 0x9d, 0x00, 0x95, 0x60, 0x9f, 0x00, 0x07, 0xb9, 0x68, 0x00, 0x09, 0x20, 0x7f, 0x00,
+    0x73, 0x78, 0x9f, 0x00, 0x01, 0x8a, 0x98, 0x00, 0xaa, 0x30, 0x99, 0x00, 0x88, 0x4c, 0x7f, 0x00,
+    0x02, 0x60, 0x7f, 0x00, 0xbc, 0x10, 0x9f, 0x00, 0x02, 0x60, 0x7f, 0x00, 0x34, 0x4a, 0x9f, 0x00,
+    0xbd, 0x90, 0x98, 0x00, 0xb1, 0x38, 0x9f, 0x00, 0x80, 0x1c, 0x7f, 0x00, 0xbc, 0x00, 0x9a, 0x00,
+    0xbc, 0x28, 0x9f, 0x00, 0xb9, 0x18, 0x9f, 0x00, 0x08, 0x60, 0x7f, 0x00, 0x01, 0x20, 0x7f, 0x00,
+    0x10, 0x24, 0x7f, 0x00, 0x4a, 0x48, 0x9f, 0x00, 0xc2, 0x38, 0x99, 0x00, 0x04, 0x1c, 0x7f, 0x00,
+    0x01, 0x40, 0x7f, 0x00, 0xcb, 0x58, 0x9f, 0x00, 0x4a, 0x48, 0x9f, 0x00, 0x45, 0x78, 0x9f, 0x00,
+    0x0a, 0x04, 0x7f, 0x00, 0x4a, 0xc8, 0x9f, 0x00, 0xd2, 0xb8, 0x99, 0x00, 0x04, 0x9c, 0x7f, 0x00,
+    0xf4, 0x98, 0x9f, 0x00, 0xd9, 0x88, 0x98, 0x00, 0x13, 0xf9, 0x9f, 0x00, 0x4a, 0xc8, 0x9f, 0x00,
+    0xd9, 0xd0, 0x9f, 0x00, 0x4a, 0x48, 0x9f, 0x00, 0x02, 0x1c, 0x7f, 0x00, 0xcb, 0x58, 0x9f, 0x00,
+    0xe8, 0x18, 0x9f, 0x00, 0x4a, 0x48, 0x9f, 0x00, 0xcb, 0x78, 0x9f, 0x00, 0x01, 0x1c, 0x7f, 0x00,
+    0x07, 0x20, 0x7f, 0x00, 0x10, 0x60, 0x7f, 0x00, 0x02, 0x20, 0x7f, 0x00, 0x4a, 0x48, 0x9f, 0x00,
+    0x45, 0x78, 0x9f, 0x00, 0x00, 0xa1, 0x7f, 0x00, 0x00, 0x80, 0x1f, 0x00, 0x00, 0xa2, 0x7f, 0x00,
+    0xd4, 0x98, 0x9f, 0x00, 0x10, 0xe0, 0x7f, 0x00, 0x02, 0xa0, 0x7f, 0x00, 0x4a, 0xc8, 0x9f, 0x00,
+    0x45, 0xf8, 0x9f, 0x00, 0x00, 0x21, 0x7f, 0x00, 0x00, 0x00, 0x1f, 0x00, 0x00, 0x22, 0x7f, 0x00,
+    0x46, 0x18, 0x9f, 0x00, 0x10, 0x60, 0x7f, 0x00, 0x02, 0x20, 0x7f, 0x00, 0x47, 0x78, 0x9f, 0x00,
+    0x18, 0x89, 0x9f, 0x00, 0x4a, 0xc8, 0x8e, 0x00, 0xd9, 0xf8, 0x9f, 0x00, 0x4a, 0xc8, 0x8d, 0x00,
+    0xd9, 0xf8, 0x9f, 0x00, 0x39, 0x11, 0x9f, 0x00, 0x1f, 0x61, 0x9f, 0x00, 0x26, 0x21, 0x8b, 0x00,
+    0x05, 0x39, 0x7f, 0x00, 0x7b, 0x78, 0x9f, 0x00, 0x05, 0x39, 0x7f, 0x00, 0x2c, 0x21, 0x9a, 0x00,
+    0x48, 0x4c, 0x7f, 0x00, 0x31, 0x19, 0x9f, 0x00, 0x08, 0x60, 0x7f, 0x00, 0x01, 0x20, 0x7f, 0x00,
+    0x08, 0x20, 0x7f, 0x00, 0x01, 0x24, 0x7f, 0x00, 0x7b, 0x78, 0x9f, 0x00, 0x39, 0x61, 0x9f, 0x00,
+    0x42, 0x21, 0x8b, 0x00, 0x05, 0x39, 0x7f, 0x00, 0x01, 0x34, 0x7f, 0x00, 0x59, 0x79, 0x9f, 0x00,
+    0x05, 0x39, 0x7f, 0x00, 0x49, 0x19, 0x9f, 0x00, 0x08, 0x60, 0x7f, 0x00, 0x01, 0x24, 0x7f, 0x00,
+    0x08, 0x20, 0x7f, 0x00, 0x01, 0x20, 0x7f, 0x00, 0x02, 0x34, 0x7f, 0x00, 0x54, 0x21, 0x9a, 0x00,
+    0x48, 0x4c, 0x7f, 0x00, 0x59, 0x79, 0x9f, 0x00, 0x01, 0x34, 0x7f, 0x00, 0x59, 0x61, 0x9f, 0x00,
+    0x01, 0x50, 0x7f, 0x00, 0x71, 0x49, 0x8b, 0x00, 0x07, 0x39, 0x7f, 0x00, 0x65, 0x19, 0x9f, 0x00,
+    0x08, 0x60, 0x7f, 0x00, 0x02, 0x24, 0x7f, 0x00, 0x02, 0x34, 0x7f, 0x00, 0x08, 0x20, 0x7f, 0x00,
+    0x8e, 0x29, 0x9a, 0x00, 0x48, 0x4c, 0x7f, 0x00, 0x8f, 0x79, 0x1f, 0x01, 0x23, 0x8a, 0x98, 0x00,
+    0x01, 0x34, 0x7f, 0x00, 0x01, 0x34, 0x7f, 0x00, 0x9f, 0x78, 0x9f, 0x00, 0x7d, 0x89, 0x98, 0x00,
+    0x00, 0x00, 0x0d, 0x00, 0x16, 0x09, 0x7f, 0x00, 0x69, 0x11, 0x7f, 0x00, 0x81, 0x19, 0x9f, 0x00,
+    0x86, 0x39, 0x9a, 0x00, 0x00, 0x18, 0x0e, 0x00, 0x07, 0x20, 0x7f, 0x00, 0x10, 0x60, 0x7f, 0x00,
+    0x01, 0x60, 0x7f, 0x00, 0x00, 0x40, 0x1f, 0x00, 0x00, 0x30, 0x7f, 0x00, 0x00, 0x28, 0x7f, 0x00,
+    0xc0, 0x68, 0x7f, 0x00, 0x01, 0x20, 0x7f, 0x00, 0xaf, 0x29, 0x9f, 0x00, 0xab, 0x11, 0x9f, 0x00,
+    0x9b, 0x61, 0x9f, 0x00, 0x07, 0x39, 0x6b, 0x00, 0x01, 0x50, 0x7f, 0x00, 0xa7, 0x29, 0x9f, 0x00,
+    0xc9, 0x49, 0x9f, 0x00, 0x9b, 0x79, 0x9f, 0x00, 0x04, 0x24, 0x7f, 0x00, 0xb7, 0x79, 0x9f, 0x00,
+    0x01, 0x34, 0x7f, 0x00, 0x9b, 0x79, 0x9f, 0x00, 0xb2, 0x31, 0x9a, 0x00, 0x48, 0x4c, 0x7f, 0x00,
+    0x04, 0x24, 0x7f, 0x00, 0xcf, 0x11, 0x9f, 0x00, 0xc0, 0x31, 0x99, 0x00, 0x88, 0x4c, 0x7f, 0x00,
+    0x02, 0x60, 0x7f, 0x00, 0xc0, 0x11, 0x9f, 0x00, 0x02, 0x60, 0x7f, 0x00, 0x01, 0x40, 0x7f, 0x00,
+    0xc7, 0x49, 0x9f, 0x00, 0xc3, 0x79, 0x9f, 0x00, 0x08, 0x24, 0x7f, 0x00, 0x01, 0x50, 0x7f, 0x00,
+    0xe9, 0x79, 0x9d, 0x00, 0x9f, 0x78, 0x9f, 0x00, 0x02, 0x34, 0x7f, 0x00, 0xb7, 0x79, 0x9f, 0x00,
+    0x01, 0x50, 0x7f, 0x00, 0xda, 0x09, 0x9f, 0x00, 0x45, 0x48, 0x8e, 0x00, 0x4b, 0x78, 0x9f, 0x00,
+    0x45, 0x48, 0x8d, 0x00, 0x4b, 0x78, 0x9f, 0x00, 0xe4, 0x09, 0x9f, 0x00, 0x00, 0x00, 0x0e, 0x00,
+    0x01, 0x00, 0x1f, 0x00, 0x00, 0x00, 0x0d, 0x00, 0x01, 0x00, 0x1f, 0x00, 0x8b, 0x68, 0x7f, 0x00,
+    0x40, 0x68, 0x7f, 0x00, 0xf3, 0x41, 0x9f, 0x00, 0xc1, 0x68, 0x7f, 0x00, 0x01, 0x00, 0x1f, 0x00,
+    0xc2, 0x68, 0x7f, 0x00, 0x01, 0x00, 0x1f, 0x00, 0x01, 0x2c, 0x7f, 0x00, 0xfd, 0x41, 0x9f, 0x00,
+    0x01, 0x00, 0x1f, 0x00, 0x05, 0x2c, 0x7f, 0x00, 0x01, 0x00, 0x1f, 0x00, 0x05, 0x39, 0x7f, 0x00,
+    0x03, 0x62, 0x9f, 0x00, 0x1b, 0x0a, 0x9f, 0x00, 0x40, 0x36, 0x7f, 0x00, 0x40, 0x35, 0x7f, 0x00,
+    0x01, 0x39, 0x7f, 0x00, 0x11, 0x4a, 0x99, 0x00, 0x2b, 0x4a, 0x9f, 0x00, 0x0d, 0x62, 0x9f, 0x00,
+    0x17, 0x4a, 0x99, 0x00, 0x2b, 0x4a, 0x9f, 0x00, 0x1d, 0x0a, 0x9f, 0x00, 0x13, 0x7a, 0x9f, 0x00,
+    0x2b, 0x4a, 0x9f, 0x00, 0x80, 0x36, 0x7f, 0x00, 0x80, 0x35, 0x7f, 0x00, 0x2b, 0x7a, 0x9f, 0x00,
+    0x05, 0x39, 0x7f, 0x00, 0x01, 0x34, 0x7f, 0x00, 0x01, 0x34, 0x7f, 0x00, 0x03, 0x7a, 0x9f, 0x00,
+    0x2b, 0x6a, 0x9f, 0x00, 0x01, 0x50, 0x7f, 0x00, 0x9f, 0x78, 0x9f, 0x00, 0x45, 0x48, 0x9f, 0x00,
+    0x4b, 0x78, 0x9f, 0x00, 0x4b, 0x90, 0x98, 0x00, 0x3b, 0x3a, 0x9f, 0x00, 0x80, 0x1c, 0x7f, 0x00,
+    0x4a, 0x00, 0x9a, 0x00, 0x4a, 0x28, 0x9f, 0x00, 0x43, 0x1a, 0x9f, 0x00, 0x08, 0x60, 0x7f, 0x00,
+    0x01, 0x20, 0x7f, 0x00, 0x10, 0x24, 0x7f, 0x00, 0x4b, 0x78, 0x9f, 0x00, 0x05, 0x20, 0x7f, 0x00,
+    0x67, 0x78, 0x9f, 0x00, 0x51, 0x8a, 0x98, 0x00, 0x00, 0x00, 0x0d, 0x00, 0x57, 0x08, 0x7f, 0x00,
+    0x29, 0x10, 0x7f, 0x00, 0x59, 0x02, 0x9f, 0x00, 0x00, 0x40, 0x1f, 0x00, 0x01, 0x60, 0x7f, 0x00,
+    0x57, 0x7a, 0x9f, 0x00, 0x24, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x24, 0x00, 0x00, 0x00,
+    0x00, 0x00, 0x00, 0x00, 0x00, 0x9a, 0x01, 0x00, 0x00, 0x00, 0x40, 0x00, 0x00, 0x00, 0xe0, 0x00,
+    0x20, 0x00, 0x01, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x03, 0x00, 0x00,
+    0x00, 0x07, 0x00, 0x00, 0x00, 0x04, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+    0x00, 0x00, 0xc0, 0x00, 0x00, 0x00, 0x00, 0x00, 0x24, 0x00, 0xc0, 0x00, 0x00, 0x0b, 0x00, 0x00,
+    0x31, 0x02, 0x02, 0x00, 0x05, 0xce, 0x00, 0x00, 0xe7, 0x60, 0x03, 0x00, 0xa8, 0xe6, 0x41, 0x09,
+    0x10, 0x05, 0x10, 0x00, 0x51, 0x1e, 0xc0, 0x42, 0x01, 0x04, 0x00, 0x00, 0x00, 0x84, 0x2e, 0x00,
+    0x1a, 0x58, 0x94, 0x16, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+    0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x01, 0x00, 0x00, 0x00,
+    0x1c, 0x03, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x0f, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+    0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+    0x06, 0x1c, 0x08, 0x00, 0x10, 0x40, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+    0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x0d, 0x00, 0x00, 0x00,
+    0x00, 0x00, 0x00, 0x00, 0x06, 0x1c, 0x08, 0x00, 0x10, 0x40, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+    0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x15, 0x24, 0xc0, 0x00,
+    0x0e, 0x00, 0x00, 0x00, 0x01, 0x00, 0x00, 0x00, 0x01, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+    0x00, 0x00, 0xc0, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+    0x00, 0x00, 0x00, 0x00, 0x55, 0x00, 0x00, 0x00, 0x00, 0x01, 0x02, 0x03, 0x04, 0x05, 0x06, 0x07,
+    0x08, 0x09, 0x0a, 0x0b, 0x0c, 0x0d, 0x0e, 0x0f, 0x10, 0x11, 0x12, 0x40, 0x41, 0x42, 0x43, 0x44,
+    0x45, 0x46, 0x47, 0x48, 0x49, 0x4a, 0x4b, 0x4c, 0x4d, 0x4e, 0x4f, 0x50, 0x51, 0x52, 0x53, 0x54,
+    0x55, 0x56, 0x57, 0x58, 0x59, 0x5a, 0x5b, 0x5c, 0x5d, 0x5e, 0x5f, 0x60, 0x61, 0x62, 0x63, 0x64,
+    0x65, 0x66, 0x67, 0x68, 0x69, 0x6a, 0x6b, 0x6c, 0x6d, 0x6e, 0x6f, 0x70, 0x71
+};
+
+static uint8_t spi_fw[] __attribute__ ((aligned (4))) = {
+    0x53, 0x45, 0x46, 0x57, 0x01, 0x00, 0x00, 0x00, 0x00, 0x00, 0x0a, 0x30, 0x01, 0x00, 0x02, 0x0b,
+    0x09, 0x00, 0xba, 0x01, 0x1c, 0x00, 0x45, 0x00, 0x18, 0x08, 0x04, 0x07, 0xce, 0x07, 0x00, 0x00,
+    0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0xa8, 0x11, 0x00,
+    0x20, 0xc0, 0x34, 0x00, 0x00, 0xc8, 0x3d, 0x00, 0x10, 0x40, 0x62, 0x00, 0x00, 0xe8, 0x67, 0x00,
+    0x30, 0xc0, 0x83, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+    0x00, 0x00, 0x00, 0x00, 0x00, 0xe8, 0xa8, 0x00, 0x30, 0x40, 0xd7, 0x00, 0x00, 0xe8, 0x11, 0x00,
+    0x10, 0xc0, 0x33, 0x00, 0x00, 0x88, 0x8a, 0x00, 0x08, 0x40, 0x00, 0x00, 0x00, 0x08, 0x8b, 0x00,
+    0x08, 0x40, 0x00, 0x00, 0x00, 0x88, 0x8c, 0x00, 0x08, 0x40, 0x90, 0x00, 0x00, 0x00, 0x00, 0x00,
+    0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+    0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+    0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x6b, 0x0b, 0x61, 0x00, 0x40, 0x11, 0x7f, 0x00,
+    0x81, 0x38, 0x7f, 0x00, 0x83, 0x7a, 0xff, 0x00, 0x5f, 0x6f, 0x7f, 0x00, 0x07, 0x78, 0x7f, 0x00,
+    0x53, 0x79, 0x7f, 0x00, 0x50, 0x78, 0x7f, 0x00, 0x9b, 0x30, 0x9f, 0x00, 0x58, 0x10, 0x9f, 0x00,
+    0x61, 0x48, 0x86, 0x00, 0x04, 0x3a, 0x7f, 0x00, 0x90, 0x68, 0x9f, 0x00, 0x6f, 0x6f, 0x7f, 0x00,
+    0x69, 0x48, 0x9f, 0x00, 0x1f, 0x78, 0x7f, 0x00, 0x83, 0x78, 0x9f, 0x00, 0x48, 0x78, 0x7f, 0x00,
+    0x0b, 0x78, 0x7f, 0x00, 0x77, 0x78, 0x9f, 0x00, 0x6f, 0x60, 0x9f, 0x00, 0x04, 0x3a, 0x64, 0x00,
+    0x72, 0x28, 0x3f, 0x01, 0x75, 0x60, 0xcc, 0x00, 0x28, 0xb8, 0x67, 0x00, 0x00, 0xba, 0x67, 0x00,
+    0x7b, 0xe8, 0x87, 0x00, 0x6e, 0xb0, 0x25, 0x01, 0x9b, 0x78, 0x9f, 0x00, 0x80, 0x28, 0xc6, 0x00,
+    0x28, 0xb8, 0x67, 0x00, 0x00, 0xb9, 0x67, 0x00, 0x87, 0xe0, 0x27, 0x01, 0x80, 0xb0, 0x85, 0x00,
+    0x9b, 0x78, 0x9f, 0x00, 0x8c, 0x28, 0x9f, 0x00, 0x00, 0x00, 0x06, 0x00, 0x28, 0xb8, 0x67, 0x00,
+    0x00, 0xba, 0x67, 0x00, 0x95, 0xe8, 0x87, 0x00, 0x91, 0xa8, 0xc5, 0x00, 0x8c, 0x30, 0x9f, 0x00,
+    0xbb, 0x38, 0x3b, 0x01, 0x0b, 0x0b, 0x63, 0x00, 0x01, 0x44, 0x7f, 0x00, 0x05, 0x39, 0x7f, 0x00,
+    0xa3, 0x60, 0x9f, 0x00, 0xb8, 0x28, 0x83, 0x00, 0xaa, 0x28, 0x83, 0x00, 0x00, 0x1e, 0x7f, 0x00,
+    0xb0, 0x28, 0x83, 0x00, 0x03, 0x39, 0x7f, 0x00, 0xaf, 0x60, 0x9f, 0x00, 0x20, 0x1c, 0x7f, 0x00,
+    0xb6, 0x30, 0x9b, 0x00, 0x40, 0x1c, 0x7f, 0x00, 0x00, 0x40, 0x1f, 0x00, 0xa7, 0x78, 0x84, 0x00,
+    0xc2, 0x48, 0x9f, 0x00, 0xc2, 0x30, 0x9a, 0x00, 0xbf, 0x60, 0x9f, 0x00, 0x04, 0x3a, 0x64, 0x00,
+    0x01, 0x44, 0x7f, 0x00, 0x20, 0x1c, 0x7f, 0x00, 0x08, 0x20, 0x7f, 0x00, 0xc9, 0x68, 0x9f, 0x00,
+    0xca, 0x08, 0x9f, 0x00, 0xff, 0x78, 0x9f, 0x00, 0xfb, 0x0b, 0x7f, 0x00, 0x89, 0x11, 0x9f, 0x00,
+    0xef, 0x50, 0x9f, 0x00, 0x01, 0x44, 0x7f, 0x00, 0x01, 0x60, 0x7f, 0x00, 0x05, 0x39, 0x7f, 0x00,
+    0xdb, 0x60, 0x9f, 0x00, 0x4b, 0x0b, 0x72, 0x00, 0x00, 0x10, 0x7f, 0x00, 0xeb, 0x18, 0x9a, 0x00,
+    0xb7, 0x08, 0x9b, 0x00, 0x01, 0x5c, 0x7f, 0x00, 0x40, 0x1c, 0x7f, 0x00, 0xb7, 0x78, 0x9f, 0x00,
+    0x00, 0x1e, 0x7f, 0x00, 0xe3, 0x78, 0x9f, 0x00, 0x50, 0xf9, 0x7f, 0x00, 0xf4, 0xd0, 0x9f, 0x00,
+    0xf3, 0xd0, 0x87, 0x00, 0xd5, 0x78, 0x85, 0x00, 0x00, 0x00, 0x01, 0x00, 0x83, 0x01, 0x9a, 0x00,
+    0x7f, 0x09, 0x9a, 0x00, 0x05, 0x39, 0x7f, 0x00, 0x05, 0x5c, 0x61, 0x00, 0x82, 0x38, 0x7f, 0x00,
+    0x09, 0x49, 0x9f, 0x00, 0x04, 0x3a, 0x7f, 0x00, 0x28, 0x69, 0x9f, 0x00, 0x08, 0x78, 0x7f, 0x00,
+    0x56, 0x79, 0x7f, 0x00, 0xef, 0x6e, 0x7f, 0x00, 0x20, 0x38, 0x7f, 0x00, 0x01, 0x2c, 0x7f, 0x00,
+    0x05, 0x2c, 0x7f, 0x00, 0x05, 0x2c, 0x7f, 0x00, 0x00, 0x04, 0x7f, 0x00, 0x01, 0x40, 0x7f, 0x00,
+    0x50, 0x78, 0x7f, 0x00, 0x02, 0x1c, 0x7f, 0x00, 0x1f, 0x61, 0x9f, 0x00, 0x25, 0x49, 0x9f, 0x00,
+    0x45, 0x79, 0x9f, 0x00, 0x48, 0x78, 0x7f, 0x00, 0x39, 0x79, 0x9f, 0x00, 0x52, 0x79, 0x7f, 0x00,
+    0xbf, 0x6e, 0x7f, 0x00, 0x2d, 0x61, 0x9f, 0x00, 0x57, 0x79, 0x9f, 0x00, 0x31, 0x61, 0x9f, 0x00,
+    0x04, 0x3a, 0x64, 0x00, 0x34, 0x29, 0x3f, 0x01, 0x37, 0x61, 0x8b, 0x00, 0x28, 0x38, 0x68, 0x00,
+    0x00, 0xba, 0x68, 0x00, 0x3d, 0xe9, 0x88, 0x00, 0x31, 0xd9, 0x28, 0x01, 0x5f, 0x79, 0x9f, 0x00,
+    0x42, 0x29, 0x9f, 0x00, 0x28, 0x38, 0x68, 0x00, 0x00, 0xb9, 0x68, 0x00, 0x49, 0xe1, 0x28, 0x01,
+    0x43, 0xd9, 0x88, 0x00, 0x5f, 0x79, 0x9f, 0x00, 0x57, 0x29, 0x88, 0x00, 0x01, 0x40, 0x7f, 0x00,
+    0x52, 0x29, 0x9f, 0x00, 0x01, 0x44, 0x7f, 0x00, 0x20, 0x38, 0x68, 0x00, 0x59, 0xd1, 0x88, 0x00,
+    0x4f, 0x59, 0x88, 0x00, 0x5f, 0x79, 0x88, 0x00, 0x0b, 0x0b, 0x7f, 0x00, 0x01, 0x44, 0x7f, 0x00,
+    0x05, 0x39, 0x7f, 0x00, 0x65, 0x61, 0x9f, 0x00, 0x7c, 0x29, 0x83, 0x00, 0x6c, 0x29, 0x83, 0x00,
+    0x00, 0x1e, 0x7f, 0x00, 0x72, 0x29, 0x83, 0x00, 0x03, 0x39, 0x7f, 0x00, 0x71, 0x61, 0x9f, 0x00,
+    0x76, 0x31, 0x9b, 0x00, 0x20, 0x1c, 0x7f, 0x00, 0x09, 0x20, 0x7f, 0x00, 0x40, 0x1c, 0x7f, 0x00,
+    0x00, 0x40, 0x1f, 0x00, 0x69, 0x79, 0x84, 0x00, 0x00, 0x1e, 0x7f, 0x00, 0xfd, 0x78, 0x9f, 0x00,
+    0x03, 0x39, 0x7f, 0x00, 0x85, 0x61, 0x9f, 0x00, 0xfb, 0x78, 0x9f, 0x00, 0x89, 0xd1, 0xdf, 0x00,
+    0x8f, 0x11, 0x9b, 0x00, 0x01, 0x5c, 0x7f, 0x00, 0x01, 0x60, 0x7f, 0x00, 0x05, 0x39, 0x7f, 0x00,
+    0x01, 0x44, 0x7f, 0x00, 0x95, 0x61, 0x9f, 0x00, 0x9b, 0x19, 0x9a, 0x00, 0x73, 0x79, 0x92, 0x00,
+    0x00, 0x1e, 0x72, 0x00, 0x73, 0x79, 0x9f, 0x00, 0x5f, 0x6e, 0x61, 0x00, 0x6b, 0x0b, 0x7f, 0x00,
+    0x40, 0x11, 0x7f, 0x00, 0x05, 0x5c, 0x7f, 0x00, 0x81, 0x38, 0x7f, 0x00, 0x83, 0x7a, 0xff, 0x00,
+    0x53, 0x79, 0x7f, 0x00, 0x09, 0x78, 0x7f, 0x00, 0x50, 0x78, 0x7f, 0x00, 0xb7, 0x49, 0x86, 0x00,
+    0x04, 0x3a, 0x7f, 0x00, 0xe6, 0x69, 0x9f, 0x00, 0x6f, 0x6e, 0x7f, 0x00, 0xbf, 0x49, 0x9f, 0x00,
+    0x1f, 0x78, 0x7f, 0x00, 0xd9, 0x79, 0x9f, 0x00, 0x48, 0x78, 0x7f, 0x00, 0x0b, 0x78, 0x7f, 0x00,
+    0xcd, 0x79, 0x9f, 0x00, 0xc5, 0x61, 0x9f, 0x00, 0x04, 0x3a, 0x64, 0x00, 0xc8, 0x29, 0x3f, 0x01,
+    0xcb, 0x61, 0xcc, 0x00, 0x28, 0xb8, 0x69, 0x00, 0x00, 0xba, 0x69, 0x00, 0xd1, 0xe9, 0x89, 0x00,
+    0xc4, 0xb1, 0x28, 0x01, 0xf1, 0x79, 0x9f, 0x00, 0xd6, 0x29, 0xc6, 0x00, 0x28, 0xb8, 0x69, 0x00,
+    0x00, 0xb9, 0x69, 0x00, 0xdd, 0xe1, 0x29, 0x01, 0xd6, 0xb1, 0x88, 0x00, 0xf1, 0x79, 0x9f, 0x00,
+    0xe2, 0x29, 0x9f, 0x00, 0x00, 0x00, 0x06, 0x00, 0x28, 0xb8, 0x69, 0x00, 0x00, 0xba, 0x69, 0x00,
+    0xeb, 0xe9, 0x89, 0x00, 0xe7, 0xa9, 0xc8, 0x00, 0xe2, 0x31, 0x9f, 0x00, 0x0b, 0x0b, 0x63, 0x00,
+    0x01, 0x44, 0x7f, 0x00, 0x05, 0x39, 0x7f, 0x00, 0xf7, 0x61, 0x9f, 0x00, 0x0c, 0x2a, 0x83, 0x00,
+    0x00, 0x00, 0x1f, 0x00, 0x00, 0x2a, 0x83, 0x00, 0x00, 0x1e, 0x7f, 0x00, 0x06, 0x2a, 0x83, 0x00,
+    0x03, 0x39, 0x7f, 0x00, 0x05, 0x62, 0x9f, 0x00, 0x20, 0x1c, 0x7f, 0x00, 0x40, 0x1c, 0x7f, 0x00,
+    0x00, 0x40, 0x1f, 0x00, 0xfd, 0x79, 0x84, 0x00, 0x23, 0x52, 0x9f, 0x00, 0x01, 0x44, 0x7f, 0x00,
+    0x01, 0x60, 0x7f, 0x00, 0x05, 0x39, 0x7f, 0x00, 0x17, 0x62, 0x9f, 0x00, 0x08, 0x5c, 0x72, 0x00,
+    0x1e, 0x1a, 0x9a, 0x00, 0x00, 0x1e, 0x7f, 0x00, 0x40, 0x1c, 0x7f, 0x00, 0x00, 0x40, 0x1f, 0x00,
+    0x50, 0xf9, 0x7f, 0x00, 0x28, 0xd2, 0x9f, 0x00, 0x27, 0xd2, 0x89, 0x00, 0x11, 0x7a, 0x88, 0x00,
+    0x00, 0x40, 0x0b, 0x00, 0x30, 0x02, 0x9f, 0x00, 0x01, 0x60, 0x7f, 0x00, 0x00, 0x40, 0x04, 0x00,
+    0x38, 0x3a, 0x9a, 0x00, 0x01, 0x40, 0x6f, 0x00, 0x37, 0x7a, 0x8f, 0x00, 0x20, 0x38, 0x6f, 0x00,
+    0x00, 0x00, 0x0f, 0x00, 0x3b, 0xd2, 0x8f, 0x00, 0x00, 0x20, 0x0f, 0x00, 0x00, 0x40, 0x1f, 0x00,
+    0x43, 0x62, 0x9f, 0x00, 0x7f, 0x6d, 0x7f, 0x00, 0x6b, 0x09, 0x7f, 0x00, 0xc4, 0x78, 0x7f, 0x00,
+    0x07, 0x78, 0x7f, 0x00, 0x04, 0x3a, 0x7f, 0x00, 0x56, 0x6a, 0x9f, 0x00, 0x01, 0x20, 0x7f, 0x00,
+    0x57, 0x7a, 0x9f, 0x00, 0x61, 0x2a, 0xc5, 0x00, 0xd5, 0x00, 0x9f, 0x00, 0x56, 0x12, 0x9f, 0x00,
+    0x04, 0x3a, 0x7f, 0x00, 0x5d, 0x6a, 0x9f, 0x00, 0x00, 0x00, 0x06, 0x00, 0x00, 0xb9, 0x67, 0x00,
+    0x82, 0x1c, 0x67, 0x00, 0x65, 0x62, 0x87, 0x00, 0x55, 0x52, 0x87, 0x00, 0x00, 0x00, 0x05, 0x00,
+    0xd5, 0x00, 0x9f, 0x00, 0x00, 0x10, 0x1f, 0x00, 0x79, 0x4a, 0x9f, 0x00, 0x04, 0x3a, 0x7f, 0x00,
+    0x73, 0x6a, 0x9f, 0x00, 0x00, 0x30, 0x05, 0x00, 0x6b, 0x7a, 0x85, 0x00, 0x6b, 0x0b, 0x7f, 0x00,
+    0x20, 0x38, 0x7f, 0x00, 0xcc, 0x78, 0x7f, 0x00, 0x9b, 0x38, 0x9b, 0x00, 0x95, 0x7a, 0x9f, 0x00,
+    0x88, 0x02, 0x9a, 0x00, 0x03, 0x39, 0x7f, 0x00, 0x87, 0x62, 0x9f, 0x00, 0x8c, 0x0a, 0x9a, 0x00,
+    0x00, 0x1e, 0x7f, 0x00, 0x05, 0x39, 0x7f, 0x00, 0x93, 0x3a, 0x9b, 0x00, 0x94, 0x0a, 0x9b, 0x00,
+    0x43, 0x2a, 0x9a, 0x00, 0x01, 0x2c, 0x7f, 0x00, 0x05, 0x2c, 0x7f, 0x00, 0x05, 0x2c, 0x7f, 0x00,
+    0x00, 0x04, 0x7f, 0x00, 0x01, 0x40, 0x7f, 0x00, 0x9f, 0x62, 0x9f, 0x00, 0x01, 0x00, 0x1f, 0x00,
+    0x6b, 0x0b, 0x61, 0x00, 0x40, 0x11, 0x7f, 0x00, 0x05, 0x5c, 0x7f, 0x00, 0x82, 0x38, 0x7f, 0x00,
+    0x83, 0x7a, 0xff, 0x00, 0x01, 0x44, 0x7f, 0x00, 0x53, 0x79, 0x7f, 0x00, 0x5f, 0x6e, 0x7f, 0x00,
+    0x09, 0x78, 0x7f, 0x00, 0x50, 0x78, 0x7f, 0x00, 0xbd, 0x4a, 0x86, 0x00, 0x04, 0x3a, 0x7f, 0x00,
+    0x24, 0x6b, 0x9f, 0x00, 0x6f, 0x6e, 0x7f, 0x00, 0xc5, 0x4a, 0x9f, 0x00, 0x1f, 0x78, 0x7f, 0x00,
+    0xfb, 0x7a, 0x9f, 0x00, 0x48, 0x78, 0x7f, 0x00, 0x0b, 0x78, 0x7f, 0x00, 0xd5, 0x7a, 0x9f, 0x00,
+    0xcb, 0x62, 0x9f, 0x00, 0x04, 0x3a, 0x64, 0x00, 0x5d, 0x03, 0x9f, 0x00, 0xce, 0x2a, 0x3f, 0x01,
+    0xd3, 0x62, 0xcc, 0x00, 0x20, 0x38, 0x69, 0x00, 0xd7, 0xd2, 0x89, 0x00, 0xea, 0x5a, 0x89, 0x00,
+    0xca, 0x32, 0x28, 0x01, 0xef, 0x6e, 0x7f, 0x00, 0x08, 0x20, 0x7f, 0x00, 0x08, 0x78, 0x7f, 0x00,
+    0x02, 0x1c, 0x7f, 0x00, 0x56, 0x79, 0x7f, 0x00, 0x01, 0x40, 0x7f, 0x00, 0x31, 0x79, 0x83, 0x00,
+    0x6f, 0x6f, 0x68, 0x00, 0x0b, 0x78, 0x7f, 0x00, 0x04, 0x20, 0x7f, 0x00, 0x01, 0x40, 0x7f, 0x00,
+    0x6e, 0x30, 0x3f, 0x01, 0x43, 0x7b, 0x9f, 0x00, 0x5d, 0x03, 0x9f, 0x00, 0xf6, 0x2a, 0xc6, 0x00,
+    0x20, 0x38, 0x69, 0x00, 0xfd, 0xd2, 0x29, 0x01, 0x12, 0x5b, 0x89, 0x00, 0xf8, 0x32, 0x88, 0x00,
+    0xef, 0x6e, 0x7f, 0x00, 0x08, 0x20, 0x7f, 0x00, 0x08, 0x78, 0x7f, 0x00, 0x02, 0x1c, 0x7f, 0x00,
+    0x56, 0x79, 0x7f, 0x00, 0x01, 0x40, 0x7f, 0x00, 0x45, 0x29, 0x83, 0x00, 0x43, 0x79, 0x9f, 0x00,
+    0x6f, 0x6f, 0x68, 0x00, 0x1f, 0x78, 0x7f, 0x00, 0x04, 0x20, 0x7f, 0x00, 0x01, 0x40, 0x7f, 0x00,
+    0x80, 0x30, 0x9f, 0x00, 0x43, 0x7b, 0x9f, 0x00, 0x5d, 0x03, 0x9f, 0x00, 0x1e, 0x2b, 0x9f, 0x00,
+    0x00, 0x00, 0x06, 0x00, 0x20, 0x38, 0x69, 0x00, 0x27, 0xd3, 0x89, 0x00, 0x36, 0x5b, 0x89, 0x00,
+    0x25, 0x2b, 0xc8, 0x00, 0x1e, 0x33, 0x9f, 0x00, 0x52, 0x79, 0x7f, 0x00, 0xbf, 0x6e, 0x63, 0x00,
+    0x57, 0x29, 0x9f, 0x00, 0x51, 0x79, 0x9f, 0x00, 0x5f, 0x6f, 0x68, 0x00, 0x07, 0x78, 0x7f, 0x00,
+    0x04, 0x20, 0x7f, 0x00, 0x01, 0x40, 0x7f, 0x00, 0x8f, 0x28, 0x9f, 0x00, 0x8c, 0x30, 0x9f, 0x00,
+    0x0b, 0x0b, 0x7f, 0x00, 0x05, 0x39, 0x7f, 0x00, 0x47, 0x63, 0x9f, 0x00, 0x4d, 0x13, 0x9a, 0x00,
+    0x00, 0x00, 0x04, 0x00, 0x50, 0x1b, 0x9a, 0x00, 0x00, 0x1e, 0x7f, 0x00, 0x56, 0x23, 0x9a, 0x00,
+    0x03, 0x39, 0x7f, 0x00, 0x55, 0x63, 0x9f, 0x00, 0x20, 0x1c, 0x7f, 0x00, 0x40, 0x1c, 0x7f, 0x00,
+    0x00, 0x40, 0x1f, 0x00, 0x62, 0x53, 0x9f, 0x00, 0x89, 0x91, 0x98, 0x00, 0xef, 0xb0, 0x98, 0x00,
+    0x01, 0x44, 0x7f, 0x00, 0x01, 0x60, 0x7f, 0x00, 0x05, 0x39, 0x7f, 0x00, 0x69, 0x63, 0x9f, 0x00,
+    0x08, 0x5c, 0x72, 0x00, 0x70, 0x1b, 0x9a, 0x00, 0x00, 0x1e, 0x7f, 0x00, 0x40, 0x1c, 0x7f, 0x00,
+    0x00, 0x40, 0x1f, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x04, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+    0x00, 0x00, 0x00, 0x00, 0x78, 0x0e, 0x24, 0x00, 0x88, 0x10, 0x01, 0x00, 0x07, 0x00, 0x24, 0x00,
+    0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x0a, 0x00, 0x01, 0x00, 0x00, 0x03, 0x00, 0x00,
+    0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x44, 0x15, 0x00,
+    0xa0, 0x83, 0x14, 0x00, 0x28, 0x81, 0xaa, 0x00, 0x02, 0x10, 0x64, 0x00, 0x00, 0x40, 0x00, 0x00,
+    0x01, 0x02, 0x00, 0x00, 0x05, 0xfe, 0x01, 0x00, 0xe7, 0xc2, 0x02, 0x00, 0x00, 0x5c, 0x43, 0x0a,
+    0x1a, 0x01, 0x10, 0x00, 0x00, 0x00, 0x80, 0x08, 0x00, 0x00, 0x00, 0x00, 0x00, 0xac, 0x0c, 0x10,
+    0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+    0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+    0xe4, 0x18, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x03, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+    0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+    0x07, 0xf8, 0x07, 0x00, 0xfe, 0xfe, 0x0f, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+    0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x20, 0x00, 0x04, 0x00, 0x00, 0x00,
+    0x09, 0x00, 0x00, 0x00, 0x07, 0xf8, 0x07, 0x00, 0xfe, 0xfe, 0x0f, 0x00, 0x00, 0x00, 0x00, 0x00,
+    0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+    0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+    0x3f, 0x03, 0xc0, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+    0x00, 0x00, 0x00, 0x00, 0x55, 0x00, 0x00, 0x00, 0x00, 0x01, 0x02, 0x03, 0x04, 0x05, 0x06, 0x07,
+    0x08, 0x09, 0x0a, 0x0b, 0x0c, 0x0d, 0x0e, 0x0f, 0x10, 0x11, 0x12, 0x40, 0x41, 0x42, 0x43, 0x44,
+    0x45, 0x46, 0x47, 0x48, 0x49, 0x4a, 0x4b, 0x4c, 0x4d, 0x4e, 0x4f, 0x50, 0x51, 0x52, 0x53, 0x54,
+    0x55, 0x56, 0x57, 0x58, 0x59, 0x5a, 0x5b, 0x5c, 0x5d, 0x5e, 0x5f, 0x60, 0x61, 0x62, 0x63, 0x64,
+    0x65, 0x66, 0x67, 0x68, 0x69, 0x6a, 0x6b, 0x6c, 0x6d, 0x6e, 0x6f, 0x70, 0x71
+};
+
+void geni_se_fw_load(struct geni_se *se, uint8_t se_mode)
+{
+	se_fw_hdr *hdr;
+	uint8_t *fw;
+	uint32_t *fw32;
+	uint32_t i, val, cfg_val_off, cfg_indx_off, cfg_limit;
+
+	if (se_mode == QUPV3_SE_UART) {
+		hdr = (se_fw_hdr*)&uart_fw;
+		fw = (uint8_t*)&uart_fw;
+	} else if (se_mode == QUPV3_SE_SPI) {
+		hdr = (se_fw_hdr*)&spi_fw;
+		fw = (uint8_t*)&spi_fw;
+	}
+	/* Disable Peripheral clock and Output control */
+	writel(0x0, se->base + SE_GENI_OUTPUT_CTRL_OFFSET);
+	writel(0x378, se->base + SE_GENI_CGC_CTRL_OFFSET);
+	writel(0x0, se->base + SE_GENI_CLK_CTRL_OFFSET);
+	writel(0x78, se->base + SE_GENI_CGC_CTRL_OFFSET);
+
+	/* Configure INIT_CFG revision */
+	val = *((uint16_t*)&hdr->config_version);
+	writel(val, se->base + SE_GENI_INIT_CFG_REVISION_OFFSET);
+
+	/* CFG_REG Programming */
+	cfg_indx_off = hdr->config_idx_offset;
+	cfg_val_off = *((uint16_t*)&hdr->config_val_offset);
+	fw32 = (uint32_t*)&fw[cfg_val_off];
+	cfg_limit = *((uint16_t*)&hdr->config_sz_in_items);
+	for (i=0; i<cfg_limit; i++) {
+		writel(fw32[i], (se->base + SE_GENI4_IMAGE_REG_OFFSET +
+				(fw[cfg_indx_off+i] * 4)));
+	}
+
+	/* If SE is not using mini-core, program CFG_RAMn region as well */
+	val = readl(se->base + SE_HW_PARAM_2_OFFSET);
+	if (!(val & SE_GENI_USE_MINICORES_MASK)) {
+		cfg_val_off = *((uint16_t*)&hdr->firmware_offset);
+		fw32 = (uint32_t*)&fw[cfg_val_off];
+		cfg_limit = *((uint16_t*)&hdr->firmware_sz_in_items);
+		memcpy((void*)(se->base + SE_GENI_CFG_RAMn_OFFSET),
+				fw32, (cfg_limit * sizeof(uint32_t)));
+	}
+
+	/* Enable Peripheral clock */
+	writel(0x1, se->base + SE_GENI_FORCE_DEFAULT_REG_OFFSET);
+	writel(0x378, se->base + SE_GENI_CGC_CTRL_OFFSET);
+	writel(0x1, se->base + SE_GENI_CLK_CTRL_OFFSET);
+	writel(0x78, se->base + SE_GENI_CGC_CTRL_OFFSET);
+
+	/* Configure FW Revision and Protocol */
+	val = ((*((uint16_t*)&hdr->serial_protocol) << 8) |
+			(*((uint16_t*)&hdr->firmware_version) >> 8));
+	writel(val, se->base + SE_GENI_FW_REVISION_OFFSET);
+	val = (*((uint16_t*)&hdr->firmware_version) & 0xFF);
+	writel(val, se->base + SE_GENI_S_FW_REVISION_OFFSET);
+
+	/* Enable Output control */
+	writel(0x7F, se->base + SE_GENI_OUTPUT_CTRL_OFFSET);
+
+	/* Enable DMA for SPI SE alone */
+	if (se_mode == QUPV3_SE_SPI)
+		writel(0x1, se->base + SE_DMA_IF_EN_OFFSET);
+}
+#endif /* CONFIG_QCOM_GENI_SE_FW_LOAD */
+
 static int geni_se_probe(struct platform_device *pdev)
 {
 	struct device *dev = &pdev->dev;
diff --git a/drivers/spi/spi-geni-qcom.c b/drivers/spi/spi-geni-qcom.c
index f4f376a8351b..f957139bdd7d 100644
--- a/drivers/spi/spi-geni-qcom.c
+++ b/drivers/spi/spi-geni-qcom.c
@@ -1087,6 +1087,10 @@ static int spi_geni_probe(struct platform_device *pdev)
 		return ret;
 	}
 
+#ifdef CONFIG_QCOM_GENI_SE_FW_LOAD
+	geni_se_fw_load(&mas->se, QUPV3_SE_SPI);
+#endif /* CONFIG_QCOM_GENI_SE_FW_LOAD */
+
 	spi->bus_num = -1;
 	spi->dev.of_node = dev->of_node;
 	spi->mode_bits = SPI_CPOL | SPI_CPHA | SPI_LOOP | SPI_CS_HIGH;
diff --git a/drivers/tty/serial/qcom_geni_serial.c b/drivers/tty/serial/qcom_geni_serial.c
index b8aa4c1293ba..76764866489c 100644
--- a/drivers/tty/serial/qcom_geni_serial.c
+++ b/drivers/tty/serial/qcom_geni_serial.c
@@ -194,6 +194,11 @@ static int qcom_geni_serial_request_port(struct uart_port *uport)
 	if (IS_ERR(uport->membase))
 		return PTR_ERR(uport->membase);
 	port->se.base = uport->membase;
+
+#ifdef CONFIG_QCOM_GENI_SE_FW_LOAD
+	geni_se_fw_load(&port->se, QUPV3_SE_UART);
+#endif /* CONFIG_QCOM_GENI_SE_FW_LOAD */
+
 	return 0;
 }
 
diff --git a/include/linux/soc/qcom/geni-se.h b/include/linux/soc/qcom/geni-se.h
index 29e06905bc1f..c0c53feb9794 100644
--- a/include/linux/soc/qcom/geni-se.h
+++ b/include/linux/soc/qcom/geni-se.h
@@ -38,6 +38,16 @@ enum geni_se_protocol_type {
 	GENI_SE_SPI_SLAVE,
 };
 
+#ifdef CONFIG_QCOM_GENI_SE_FW_LOAD
+enum {
+	QUPV3_SE_UART = 0,
+	QUPV3_SE_HSUART,
+	QUPV3_SE_SPI,
+	QUPV3_SE_I2C,
+	QUPV3_SE_MAX,
+};
+#endif
+
 struct geni_wrapper;
 struct clk;
 
@@ -521,5 +531,9 @@ void geni_icc_set_tag(struct geni_se *se, u32 tag);
 int geni_icc_enable(struct geni_se *se);
 
 int geni_icc_disable(struct geni_se *se);
+
+#ifdef CONFIG_QCOM_GENI_SE_FW_LOAD
+void geni_se_fw_load(struct geni_se *se, uint8_t se_mode);
+#endif
 #endif
 #endif
-- 
2.34.1

